### PR TITLE
Integration of Vesting Pallet into the Runtime and Addition of Related Tests and Dependencies

### PR DESCRIPTION
**Pull request by asiniscalchi**
## **Type**
Enhancement

___

## **Description**
- The Vesting Pallet has been integrated into the runtime, with configuration parameters set.
- Tests have been added to check the configuration of the Vesting Pallet and simulate vesting over time.
- The Vesting Pallet has been added as a dependency in the main `Cargo.toml` file and as a workspace dependency in the `ownership-chain/runtime/Cargo.toml` file.
- The Vesting Pallet has been included in the list of packages for std, runtime-benchmarks, and try-runtime features in the `ownership-chain/runtime/Cargo.toml` file.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Integration of Vesting Pallet into the Runtime</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/src/lib.rs
- Added configuration for the Vesting Pallet.
- Integrated the Vesting Pallet into the runtime.
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/295/files#diff-b0f69a71ca20f9dbda0917139a7a931cd6503375c8833428c8ac431ea693aef4">+21/-3</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>constant_tests.rs</strong><dd><code>Addition of Tests for Vesting Pallet</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/src/tests/constant_tests.rs
- Added tests to check the configuration of the Vesting Pallet.
- Added a test to simulate vesting over time.
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/295/files#diff-9ca87a6e5886f66b923cf3a0efbbda0c1f422b9aef9d200a65ce525e47aed69c">+51/-2</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Addition of Vesting Pallet Dependency</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Cargo.toml
- Added the Vesting Pallet as a dependency.
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/295/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Update of Workspace Dependencies and Features</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ownership-chain/runtime/Cargo.toml
- Added the Vesting Pallet as a workspace dependency.
- Included the Vesting Pallet in the list of packages for std, <br>  runtime-benchmarks, and try-runtime features.
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/295/files#diff-a6f1e3deb5d45fdcea9465ba45b4c048309ff47223bd15d61c1fe1340e98200f">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Review by dastanfv — DISMISSED** [review]

File: ownership-chain/runtime/src/tests/constant_tests.rs [hunk 1]
```diff
@@ -56,3 +57,51 @@ fn send_1_minimum_unit_to_wallet_with_0_wei_balance_should_increase_balance_by_1
...
+	assert_eq!(<Runtime as pallet_vesting::Config>::MAX_VESTING_SCHEDULES, 28);
+}
+
+#[test]
+fn account_vests_correctly_over_time() {
+	new_test_ext().execute_with(|| {
+		let alice = AccountId::from_str(ALICE).unwrap();
+		let bob = AccountId::from_str(BOB).unwrap();
+		let cliff_duration = 24_u32;
+		let vesting_duration = 24 * 4; // 4 times the cliff duration
```
Comment (dastanfv) [inline ownership-chain/runtime/src/tests/constant_tests.rs:77]:
```suggestion
		let vesting_duration = cliff_duration * 4; // 4 times the cliff duration
```

File: ownership-chain/runtime/src/tests/constant_tests.rs [hunk 2]
```diff
@@ -56,3 +57,51 @@ fn send_1_minimum_unit_to_wallet_with_0_wei_balance_should_increase_balance_by_1
...
+		let per_block = UNIT; // Amount vested per block
```
Comment (dastanfv) [inline ownership-chain/runtime/src/tests/constant_tests.rs:78]:
```suggestion
		let amount_per_block = UNIT; // Amount vested per block
```

**Review by dastanfv — DISMISSED** [review]

**Review by magecnion — CHANGES_REQUESTED** [review]

File: ownership-chain/runtime/src/tests/constant_tests.rs [hunk 3]
```diff
@@ -56,3 +57,51 @@ fn send_1_minimum_unit_to_wallet_with_0_wei_balance_should_increase_balance_by_1
...
+
+		// Transfer vested funds from Alice to Bob
+		assert_ok!(Vesting::vested_transfer(
+			RuntimeOrigin::signed(alice),
+			bob.clone(),
+			vesting_info
+		));
+
+		// Simulate block progression and check Bob's balance each block
+		for block_num in cliff_duration..=cliff_duration + vesting_duration as u32 {
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply ownership-chain/runtime/src/tests/constant_tests.rs:99]: first step of the loop check that

File: ownership-chain/runtime/src/tests/constant_tests.rs [hunk 4]
```diff
@@ -56,3 +57,51 @@ fn send_1_minimum_unit_to_wallet_with_0_wei_balance_should_increase_balance_by_1
...
+			frame_system::Pallet::<Runtime>::set_block_number(block_num);
+			assert_ok!(Vesting::vest(RuntimeOrigin::signed(bob.clone())));
+			let vested_amount = (block_num - cliff_duration) as u128 * amount_per_block;
+			assert_eq!(Balances::usable_balance(&bob), vested_amount);
+			assert_eq!(Balances::total_balance(&bob), total_vested_amount);
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply ownership-chain/runtime/src/tests/constant_tests.rs:104]: I agree is a superfluous check

File: ownership-chain/runtime/src/tests/constant_tests.rs [hunk 5]
```diff
@@ -56,3 +57,51 @@ fn send_1_minimum_unit_to_wallet_with_0_wei_balance_should_increase_balance_by_1
...
+		<Runtime as pallet_vesting::Config>::UnvestedFundsAllowedWithdrawReasons::get(),
+		WithdrawReasons::except(WithdrawReasons::TRANSFER | WithdrawReasons::RESERVE)
+	);
+	assert_eq!(<Runtime as pallet_vesting::Config>::MAX_VESTING_SCHEDULES, 28);
+}
+
+#[test]
+fn account_vests_correctly_over_time() {
+	new_test_ext().execute_with(|| {
+		let alice = AccountId::from_str(ALICE).unwrap();
```
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply ownership-chain/runtime/src/tests/constant_tests.rs:74]: don't think it is needed. I am checking the functionality here.

File: ownership-chain/runtime/src/tests/constant_tests.rs
[the remark this answers is not in the document]
Comment (asiniscalchi) [reply ownership-chain/runtime/src/tests/constant_tests.rs:1]: https://github.com/freeverseio/laos/pull/298 <--- this PR moves the tests into mod.rs

**Comment by dastanfv** [comment]
just saw that the file of the tests is called `constant_tests.rs`. I agree with @magecnion that it should be renamed, possibly to `vesting_tests.rs` or sth like that.

**Review by magecnion — APPROVED** [review]

**Review by magecnion — APPROVED** [review]

**Review by dastanfv — APPROVED** [review]

**Pull request merged**